src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -0,0 +1,275 @@
+    /// </summary>
+    internal sealed class FileChangeWatcher
+    {
+        /// <summary>
+        /// Gate that is used to guard modifications to <see cref="_taskQueue"/>.
+        /// </summary>
+        private readonly object _taskQueueGate = new object();
+
+        /// <summary>
+        /// We create a queue of tasks against the IVsFileChangeEx service for two reasons. First, we are obtaining the service asynchronously, and don't want to
+        /// block on it being available, so anybody who wants to do anything must wait for it. Secondly, the service itself is single-threaded; in the past
+        /// we've blocked up a bunch of threads all trying to use it at once. If the latter ever changes, we probably want to reconsider the implementation of this.
+        /// For performance and correctness reasons, NOTHING should ever do a block on this; figure out how to do your work without a block and add any work to
+        /// the end of the queue.
+        /// </summary>
+        private Task<IVsFileChangeEx> _taskQueue;
+        private readonly static Func<Task<IVsFileChangeEx>, object, IVsFileChangeEx> _executeActionDelegate =
+            (precedingTask, state) => { ((Action<IVsFileChangeEx>)state)(precedingTask.Result); return precedingTask.Result; };
+
+        public FileChangeWatcher(Task<IVsFileChangeEx> fileChangeService)
+        {
+            _taskQueue = fileChangeService;
+        }
+
+        private void EnqueueWork(Action<IVsFileChangeEx> action)
+        {
+            lock (_taskQueueGate)
+            {
+                _taskQueue = _taskQueue.ContinueWith(
+                    _executeActionDelegate,
+                    action,
+                    CancellationToken.None,
+                    TaskContinuationOptions.None,
+                    TaskScheduler.Default);
+            }
+        }
+
+        // TODO: remove this when there is a mechanism for a caller of EnqueueWatchingFile
+        // to explicitly wait on that being complete.
+        public void WaitForQueue_TestOnly()
+        {
+            Task queue;
+
+            lock (_taskQueueGate)
+            {
+                queue = _taskQueue;
+            }
+
+            queue.Wait();
+        }
+
+        public IContext CreateContext()
+        {
+            return new Context(this, null);
+        }
+
+        /// <summary>
+        /// Creates an <see cref="IContext"/> that watches all files in a directory, in addition to any files explicitly requested by <see cref="IContext.EnqueueWatchingFile(string)"/>.
+        /// </summary>
+        public IContext CreateContextForDirectory(string directoryFilePath)
+        {
+            if (directoryFilePath == null)
+            {
+                throw new ArgumentNullException(nameof(directoryFilePath));
+            }
+
+            return new Context(this, directoryFilePath);
+        }
+
+        public interface IContext : IDisposable
```

Comment:
>IContext [](start = 25, length = 8)

Why this interface? Is this implemented by other than Context class?

Reply:
Comment added -- if I make the type public, it leaks implementation details I want to keep private.